fix: handle uninitialized and non-leader states in Raft cluster mode

chanfun-ren opened the pull request:
#307 
read/write commands are now dropped if raft has not been initialized in raft mode, and an error message 'NOCLUSTER No Raft Cluster' is returned to the client. This change prevents the system from crashing due to failed assertions.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **新功能**
  - 在执行命令之前添加了PRAFT初始化和领导状态检查逻辑。

- **测试**
  - 修改了与数据库刷新相关的测试用例，以增加对特定错误消息的验证逻辑。
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

chanfun-ren committed:
fix: prevent crash by rejecting r/w cmd pre-raft initialization (#307)

read/write commands are now dropped if raft has not been initialized
in raft mode, and an error message 'NOCLUSTER No Raft Cluster' is
returned to the client. This change prevents the system from crashing
due to failed assertions.

LuckyTom1024 commented on src/base_cmd.cc:
```diff
@@ -66,6 +73,16 @@ std::string BaseCmd::Name() const { return name_; }
 
+bool BaseCmd::IsAllowedPreRaftInit(const std::string& cmd_name) const {
+  // Allow all commands if Raft is not used or if Raft is initialized
+  if (!g_config.use_raft.load() || PRAFT.IsInitialized()) {
```
这里的 use_raft 指的是是否构建 raft 集群模式，true：构建 raft 集群模式，false：构建主从模式，但 PikiwiDB 的主从模式也是利用的 raft 来做日志管理，只是从节点的角色是 learner，不参与实际的日志投票。所以这里只需要判断是否初始化了即可

chanfun-ren replied:
ok

chanfun-ren replied:
> 这里的 use_raft 指的是是否构建 raft 集群模式，true：构建 raft 集群模式，false：构建主从模式，但 PikiwiDB 的主从模式也是利用的 raft 来做日志管理，只是从节点的角色是 learner，不参与实际的日志投票。所以这里只需要判断是否初始化了即可

ok

chanfun-ren replied:
后续：由于 pikiwidb 目前 `主从模式也是利用的 raft 来做日志管理` 这部分功能暂未实现，所以这个 pr 还是先对 use_raft 是否为 true 进行判断。（后续实现了可以在这个基础上再改）

chanfun-ren committed:
Remove `use_raft` dependency in IsAllowedPreRaftInit() checks

Adjust the command checks in IsAllowedPreRaftInit by removing the
dependency on the use_raft flag, since even in PikiwiDB's master-slave
mode, Raft is used for log management.

chanfun-ren committed:
fix: adjust test

longfar-ncy commented on src/base_cmd.h:
```diff
@@ -148,6 +149,10 @@ const std::string kCmdNameZRevrank = "zrevrank";
 const std::string kCmdNameZIncrby = "zincrby";
 
+// If Raft is used but not initialized, only allow specific commands
+const std::set<std::string> kPreRaftInitCmds = {kCmdNameRaftCluster, kCmdNamePing, kCmdNameConfig,
```
看能不能不使用 set 的方式，通过命令是否有 flags 判断

chanfun-ren replied:
ok

chanfun-ren replied:
- 之前 set 的实现方式是会允许未初始化 raft 时候 info, config 等命令的执行。
- 我看了下目前的代码，通过 flag 来实现可能会不太符合预期。比如 pikiwidb 的 InfoCmd 加上了 kCmdFlagsReadonly flag 和 kAclCategoryAdmin , 但是看 redis 那边 info cmd 没有被归为 read 类/管理类的命令，如下图。类似的命令还有 ping, shutdown, config 和 redis 那边对不上。
- 我理解这个要改的话 应该先把那些命令 flag 和 Category ”纠正“过来？（我不太确定 pikiwi 命令 flag 和 redis 那边不同步是否有额外的考量），然后根据命令是否有读写 flag/category 判断是否允许执行。
1. pikiwi `info`:
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/55045688/a7cf28ae-93de-4a48-b45f-3d8e10c53b88)

2. redis `info`:
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/55045688/7467ae49-570a-4b59-b87d-35449cf86e4e)

chanfun-ren replied:
后续：尽管 pikiwi 目前实现的部分命令 read write flag 不太合适，但是这个 pr 的功能仍然根据命令的 r/w flag 来实现。后续会逐步修正 pikiwi 命令的 flag.

longfar-ncy commented:
https://github.com/OpenAtomFoundation/pikiwidb/pull/302 帮忙把这个内容加进这个pr

chanfun-ren committed:
Handle uninitialized and non-leader states in Raft cluster mode

1. Return an error message when PRAFT is not initialized and a read/write command is received.
2. Redirect write commands if the server is not the leader.

chanfun-ren commented:
> #302 帮忙把这个内容加进这个pr

已加。

对于处于 raft 集群模式下的请求：
  1. 如果 PRAFT 还没有初始化，对于读写命令，返回客户端一个初始化未完成的提示信息。#307
  2. 如果 PRAFT 初始化完成了，对于写命令，如果自身不是 leader，返回一个重定向的信息。 #302

luky116 commented:
@开颜 @changyuan

AlexStocks commented:
@CodiumAI-Agent  /improve

QodoAI-Agent commented:
## PR Code Suggestions ✨

<table><thead><tr><td>Category</td><td align=left>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </td><td align=center>Score</td></tr><tbody><tr><td rowspan=1><strong>Possible bug</strong></td>
<td>



<details><summary>Add a null check for <code>PRAFT</code> to prevent potential null pointer dereferences</summary>

___


**Consider checking if <code>PRAFT</code> is not null before calling its methods to avoid potential null <br>pointer dereferences.**
    
[src/base_cmd.cc [39-41]](https://github.com/OpenAtomFoundation/pikiwidb/pull/316/files#diff-23c5847b03356e8dd665fe88eaea18e5a2ba197586895ecc56426ea9186f7fafR39-R41)

```diff
-if (!PRAFT.IsInitialized() && (HasFlag(kCmdFlagsReadonly) || HasFlag(kCmdFlagsWrite))) {
+if (PRAFT && !PRAFT.IsInitialized() && (HasFlag(kCmdFlagsReadonly) || HasFlag(kCmdFlagsWrite))) {
   DEBUG("drop command: {}", client->CmdName());
   return client->SetRes(CmdRes::kErrOther, "PRAFT is not initialized");
 }
 
```
                   


<details><summary><b>Suggestion importance[1-10]: 9</b></summary>

Why: Adding a null check for `PRAFT` is a crucial improvement to prevent potential null pointer dereferences, which could lead to runtime crashes.


</details></details></td><td align=center>9

</td></tr><tr><td rowspan=1><strong>Enhancement</strong></td>
<td>



<details><summary>Use more descriptive error codes for specific failure scenarios</summary>

___


**Consider using a more specific error code than <code>kErrOther</code> for uninitialized <code>PRAFT</code> and <br>non-leader scenarios to improve error handling and client debugging.**
    
[src/base_cmd.cc [41-46]](https://github.com/OpenAtomFoundation/pikiwidb/pull/316/files#diff-23c5847b03356e8dd665fe88eaea18e5a2ba197586895ecc56426ea9186f7fafR41-R46)

```diff
-return client->SetRes(CmdRes::kErrOther, "PRAFT is not initialized");
-return client->SetRes(CmdRes::kErrOther, fmt::format("MOVED {}", PRAFT.GetLeaderAddress()));
+return client->SetRes(CmdRes::kErrPraftUninitialized, "PRAFT is not initialized");
+return client->SetRes(CmdRes::kErrNotLeader, fmt::format("MOVED {}", PRAFT.GetLeaderAddress()));
 
```
                   


<details><summary><b>Suggestion importance[1-10]: 8</b></summary>

Why: Using more descriptive error codes enhances error handling and client debugging, making it easier to diagnose specific issues.


</details></details></td><td align=center>8

</td></tr><tr><td rowspan=1><strong>Performance</strong></td>
<td>



<details><summary>Optimize the atomic load operation with relaxed memory ordering for potential performance improvement</summary>

___


**Use <code>std::atomic<bool>::load(std::memory_order_relaxed)</code> for potentially better performance when <br>the ordering of use_raft is not critical.**
    
[src/base_cmd.cc [37]](https://github.com/OpenAtomFoundation/pikiwidb/pull/316/files#diff-23c5847b03356e8dd665fe88eaea18e5a2ba197586895ecc56426ea9186f7fafR37-R37)

```diff
-if (g_config.use_raft.load()) {
+if (g_config.use_raft.load(std::memory_order_relaxed)) {
 
```
                   


<details><summary><b>Suggestion importance[1-10]: 7</b></summary>

Why: This suggestion is correct and can improve performance, but the impact is minor as it only optimizes a single atomic load operation.


</details></details></td><td align=center>7

</td></tr><tr><td rowspan=1><strong>Maintainability</strong></td>
<td>



<details><summary>Refactor complex condition checks into a separate method to enhance readability and maintainability</summary>

___


**Refactor the condition checks into a separate method to improve code readability and <br>maintainability.**
    
[src/base_cmd.cc [39-46]](https://github.com/OpenAtomFoundation/pikiwidb/pull/316/files#diff-23c5847b03356e8dd665fe88eaea18e5a2ba197586895ecc56426ea9186f7fafR39-R46)

```diff
-if (!PRAFT.IsInitialized() && (HasFlag(kCmdFlagsReadonly) || HasFlag(kCmdFlagsWrite))) {
-  DEBUG("drop command: {}", client->CmdName());
-  return client->SetRes(CmdRes::kErrOther, "PRAFT is not initialized");
+if (checkPraftState(client)) {
+  return;
 }
-if (!PRAFT.IsLeader() && HasFlag(kCmdFlagsWrite)) {
-  return client->SetRes(CmdRes::kErrOther, fmt::format("MOVED {}", PRAFT.GetLeaderAddress()));
-}
+// Define checkPraftState elsewhere in the class
 
```
                   


<details><summary><b>Suggestion importance[1-10]: 6</b></summary>

Why: Refactoring condition checks into a separate method improves readability and maintainability, but it is a minor enhancement compared to functional or performance improvements.


</details></details></td><td align=center>6

</td></tr></tr></tbody></table>

AlexStocks commented:
@CodiumAI-Agent /review

QodoAI-Agent commented:
## PR Review 🔍

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

3, because the PR involves changes in the core command execution logic and introduces new conditions based on the state of the Raft cluster. Understanding the implications of these changes on the overall system behavior and ensuring they do not introduce new bugs or regressions requires a moderate level of effort.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

Yes


</td></tr>
<tr><td> ⚡&nbsp;<strong>Possible issues</strong></td><td>

<strong>Possible Bug:</strong> The new code in `BaseCmd::Execute` does not handle the case where a command is neither read-only nor write, which might lead to unexpected behaviors for such commands.

</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>src/base_cmd.cc
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider adding an else clause to handle commands that do not fall into read-only or write categories. This will ensure that all command types are accounted for and handled appropriately. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/OpenAtomFoundation/pikiwidb/pull/316/files#diff-23c5847b03356e8dd665fe88eaea18e5a2ba197586895ecc56426ea9186f7fafR46'>return client->SetRes(CmdRes::kErrOther, fmt::format("MOVED {}", PRAFT.GetLeaderAddress()));</a></td></tr></table><hr>

</details>

longfar-ncy commented on src/base_cmd.cc:
```diff
@@ -32,6 +34,19 @@ std::vector<std::string> BaseCmd::CurrentKey(PClient* client) const { return std
+    }
+
+    // 2. 如果 PRAFT 初始化完成了，对于写命令，如果自身的角色不是 leader，返回一个重定向的信息。
+    if (!PRAFT.IsLeader() && HasFlag(kCmdFlagsWrite)) {
```
先判断 flag 吧，如果能短路提前退出最好，leader的判断要进入 braft，braft还加了锁

chanfun-ren replied:
done

longfar-ncy commented on src/base_cmd.cc:
```diff
@@ -32,6 +34,19 @@ std::vector<std::string> BaseCmd::CurrentKey(PClient* client) const { return std
   DEBUG("execute command: {}", client->CmdName());
 
+  if (g_config.use_raft.load()) {
+    // 1. 如果 PRAFT 还没有初始化，对于读写命令，应该返回客户端一个初始化未完成的提示信息。
```
注释麻烦改成英文叭

chanfun-ren replied:
done

chanfun-ren committed:
optimize conditional checks and update comments to English

1. Updated the conditional logic to prioritize command flag checks
   before leader status. This enables early exit and reduces the
   overhead associated with entering the braft leader check,
   which involves locking mechanisms.

2. Translated existing Chinese comments into English.

longfar-ncy commented on tests/consistency_test.go:
```diff
@@ -47,11 +47,12 @@ var _ = Describe("Consistency", Ordered, func() {
 				c := s.NewClient()
 				Expect(c).NotTo(BeNil())
-				Expect(c.FlushDB(ctx).Err()).NotTo(HaveOccurred())
+				// Expect(c.FlushDB(ctx).Err()).NotTo(HaveOccurred())
```
这行注释删掉吧

chanfun-ren replied:
done

longfar-ncy commented on tests/consistency_test.go:
```diff
@@ -47,11 +47,12 @@ var _ = Describe("Consistency", Ordered, func() {
 				leader = s.NewClient()
 				Expect(leader).NotTo(BeNil())
-				Expect(leader.FlushDB(ctx).Err()).NotTo(HaveOccurred())
+				// Expect(leader.FlushDB(ctx).Err()).NotTo(HaveOccurred())
```
这行要不也换成和 55 行一样的？

chanfun-ren replied:
done

chanfun-ren committed:
adjust consistency test code.
